map: Add lookup_into() method

Jongy opened the pull request:
Allows looking up into a preallocated buffer.

This API is just a proposal - lmk what you think?

Jongy commented on libbpf-rs/src/map.rs:
```diff
@@ -483,6 +507,32 @@ pub trait MapCore: Debug + AsFd + private::Sealed {
+    fn lookup_into(&self, key: &[u8], value: &mut [u8], flags: MapFlags) -> Result<bool> {
+        check_not_bloom_or_percpu(self)?;
+
+        if value.len() != self.value_size() as usize {
```
size check is done here and not in `lookup_raw_ptr` since `lookup_raw` does not require it

d-e-s-o commented:
> I'm not sure why the tests fail on nightly/dev

It's the debug assertion you added. It doesn't work for the per-CPU case.

Jongy commented:
> > I'm not sure why the tests fail on nightly/dev
> 
> It's the debug assertion you added. It doesn't work for the per-CPU case.

Haha nice. Will run the full test suite next time.
I'll find another place for this assert then

Jongy commented:
I removed the SAFETY from the non-unsafe function. I did keep the `debug_assert_eq`, just fixed the condition, because I think it's a helpful guard for this internal function, and it has zero cost in release builds. Practically I don't expect it to hit any of our users (in their debug builds) but if we're unsure, we can gate it with `cfg(test)`?
